### DISCUSSION TITLE
Poor performance of wasm binary execution over remote REST api call of OPA server.

### DISCUSSION THREAD

**Discussion by rai-prashanna** [open-policy-agent/community#440] [Wasm]
I notice that when experimenting with wasm binary, evaluation of input against wasm binary was very poor in comparison to the remote call of REST API of OPA server. 
I experimented with wasm binary, JARL library, OPA rest API server.
I created bundle with optimization flag=1 i.e o=1 and run as server. 

Both plan.json and wasm files are located locally
Despite OPA being run as a REST API server in a separate container with an overhead of sending and processing requests/responses.

![poor-performance](https://github.com/open-policy-agent/community/assets/21035884/7208c1f2-9608-4fb0-99a1-46c335e5a1b7)

The green line is used for OPA rest,
The violent line is used for WASM,
The purple line is used fro JARL.

My policy can be found in this link.
https://play.openpolicyagent.org/p/W7dh9TIqMi
https://play.openpolicyagent.org/p/LYgbgZj0Qg

My link to the repo of testing
https://github.com/rai-prashanna/performanceTesting

The performance of the wasm doesn't match with the details mentioned in this document.
https://www.openpolicyagent.org/docs/latest/integration/

![image](https://github.com/open-policy-agent/community/assets/21035884/8b2dd413-e0e0-474b-93f5-fe558fee15ad)

**Comment by srenatus** [comment]
Wasm performance depends on the runtime that's used, too. And if that runtime is using the 1.2 ABI or not. You could try to figure that out, I suppose.

**Reply by rai-prashanna** [reply]
so what is the best version of runtime that should be used, if I want to execute in Java applications?

**Reply by srenatus** [reply]
Sorry -- I don't know 😅
